smart-contracts/osmosis/contracts/cl-vault/src/vault/swap.rs

```diff
@@ -30,51 +28,77 @@ pub fn execute_swap_non_vault_funds(
     let mut swap_msgs: Vec<CosmosMsg> = vec![];
 
     for swap_operation in swap_operations {
-        let token_in_denom = swap_operation.token_in_denom.clone();
-        let pool_token_0 = pool_config.token0.clone();
-        let pool_token_1 = pool_config.token1.clone();
+        let token_in_denom = &swap_operation.token_in_denom;
 
-        if token_in_denom == pool_token_0 || token_in_denom == pool_token_1 {
+        if token_in_denom == &pool_config.token0 || token_in_denom == &pool_config.token1 {
             return Err(ContractError::InvalidSwapAssets {});
         }
 
-        let balance_in_contract = deps
+        let token_in_balance = deps
             .querier
-            .query_balance(
-                contract_address.clone(),
-                swap_operation.clone().token_in_denom,
-            )?
+            .query_balance(env.clone().contract.address, token_in_denom.clone())?
             .amount;
-
-        if balance_in_contract == Uint128::zero() {
-            return Err(ContractError::InsufficientFunds {});
+        if token_in_balance.is_zero() {
+            return Err(ContractError::InsufficientFundsForSwap {
+                balance: token_in_balance,
+                needed: Uint128::new(1),
+            });
         }
 
-        let part_0_amount = balance_in_contract.checked_div(Uint128::new(2))?;
-        let part_1_amount = balance_in_contract
-            .checked_add(Uint128::new(1))?
-            .checked_div(Uint128::new(2))?;
-
-        let token_out_min_amount_0 =
-            part_0_amount.checked_mul_floor(vault_config.swap_max_slippage)?;
-        let token_out_min_amount_1 =
-            part_1_amount.checked_mul_floor(vault_config.swap_max_slippage)?;
+        let (base_ratio, quote_ratio) = get_asset_ratios(deps.storage, &deps.querier)?;
+        let to_token0_amount = Decimal::from_ratio(token_in_balance, Uint128::one())
+            .checked_mul(base_ratio)?
+            .to_uint_floor();
+        let to_token1_amount = Decimal::from_ratio(token_in_balance, Uint128::one())
+            .checked_mul(quote_ratio)?
+            .to_uint_floor();
 
         let dex_router = DEX_ROUTER.may_load(deps.storage)?;
+
+        let twap_price_base = get_twap_price(
```

Comment:
That is twice the same code for base and quote. Is it possible to extract that to a function that returns the swap_msg?

Reply:
Addressed by creating preapare_swap_msg helper. This could be further generazlied in the future if we want to use it also on deposit.rs and range.rs as the behavior are very similar.